File: src/app/lib/models/event_dto.dart
import 'buff.dart';
import 'event_type.dart';
import 'pokemon.dart';

/// Primary DTO for a Pokemon GO event, with all data pre-shaped for display.
class EventDto {
  final String id;
  final String name;
  final EventType eventType;
  final String heading;
  final String imageUrl;
  final String linkUrl;
  final DateTime? start;
  final DateTime? end;
  final bool isUtcTime;
  final bool hasSpawns;
  final bool hasResearchTasks;
  final List<Buff> buffs;
  final List<Pokemon> featuredPokemon;
  final List<String> promoCodes;

  const EventDto({
    required this.id,
    required this.name,
    required this.eventType,
    required this.heading,
    required this.imageUrl,
    required this.linkUrl,
    this.start,
    this.end,
    required this.isUtcTime,
    required this.hasSpawns,
    required this.hasResearchTasks,
    required this.buffs,
    required this.featuredPokemon,
    required this.promoCodes,
  });

  factory EventDto.fromJson(Map<String, dynamic> json) => EventDto(
        id: json['id'] as String,
        name: json['name'] as String,
        eventType: EventType.fromJson(json['eventType'] as String),
        heading: json['heading'] as String,
        imageUrl: json['imageUrl'] as String,
        linkUrl: json['linkUrl'] as String,
        start: json['start'] != null
            ? DateTime.parse(json['start'] as String)
            : null,
        end: json['end'] != null
            ? DateTime.parse(json['end'] as String)
            : null,
        isUtcTime: json['isUtcTime'] as bool,
        hasSpawns: json['hasSpawns'] as bool,
        hasResearchTasks: json['hasResearchTasks'] as bool,
        buffs: (json['buffs'] as List<dynamic>)
            .map((e) => Buff.fromJson(e as Map<String, dynamic>))
            .toList(),
        featuredPokemon: (json['featuredPokemon'] as List<dynamic>)
            .map((e) => Pokemon.fromJson(e as Map<String, dynamic>))
            .toList(),
        promoCodes: (json['promoCodes'] as List<dynamic>)
            .map((e) => e as String)
            .toList(),
      );

  Map<String, dynamic> toJson() => {
        'id': id,
        'name': name,
        'eventType': eventType.toJson(),
        'heading': heading,
        'imageUrl': imageUrl,
        'linkUrl': linkUrl,
        if (start != null) 'start': start!.toIso8601String(),
        if (end != null) 'end': end!.toIso8601String(),
        'isUtcTime': isUtcTime,
        'hasSpawns': hasSpawns,
        'hasResearchTasks': hasResearchTasks,
        'buffs': buffs.map((e) => e.toJson()).toList(),
        'featuredPokemon': featuredPokemon.map((e) => e.toJson()).toList(),
        'promoCodes': promoCodes,
      };
}

File: src/backend/GoCalGo.Contracts/Events/EventDto.cs
using System.Text.Json.Serialization;

namespace GoCalGo.Contracts.Events
{
    /// <summary>
    /// Primary DTO for a Pokemon GO event, with all data pre-shaped for display.
    /// Produced by the backend from ScrapedDuck data; consumed by the Flutter app.
    /// </summary>
    public sealed record EventDto
    {
        /// <summary>Unique identifier for the event (e.g. "community-day-2026-03").</summary>
        [JsonPropertyName("id")]
        public required string Id { get; init; }

        /// <summary>Display name of the event.</summary>
        [JsonPropertyName("name")]
        public required string Name { get; init; }

        /// <summary>Categorised event type (e.g. community-day, spotlight-hour).</summary>
        [JsonPropertyName("eventType")]
        public required EventTypeDto EventType { get; init; }

        /// <summary>Short subtitle or tagline for the event.</summary>
        [JsonPropertyName("heading")]
        public required string Heading { get; init; }

        /// <summary>URL to the event's banner or promotional image.</summary>
        [JsonPropertyName("imageUrl")]
        public required string ImageUrl { get; init; }

        /// <summary>URL to the official event page on pokemongolive.com.</summary>
        [JsonPropertyName("linkUrl")]
        public required string LinkUrl { get; init; }

        /// <summary>Event start time. Null if the start time is unknown.</summary>
        [JsonPropertyName("start")]
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
        public DateTime? Start { get; init; }

        /// <summary>Event end time. Null if the end time is unknown.</summary>
        [JsonPropertyName("end")]
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
        public DateTime? End { get; init; }

        /// <summary>Whether start/end times are in UTC (true) or local to the player (false).</summary>
        [JsonPropertyName("isUtcTime")]
        public required bool IsUtcTime { get; init; }

        /// <summary>Whether the event features increased wild Pokemon spawns.</summary>
        [JsonPropertyName("hasSpawns")]
        public required bool HasSpawns { get; init; }

        /// <summary>Whether the event includes special field research tasks.</summary>
        [JsonPropertyName("hasResearchTasks")]
        public required bool HasResearchTasks { get; init; }

        /// <summary>Active buffs and bonuses during this event.</summary>
        [JsonPropertyName("buffs")]
        public required IReadOnlyList<BuffDto> Buffs { get; init; }

        /// <summary>Pokemon highlighted or featured in this event.</summary>
        [JsonPropertyName("featuredPokemon")]
        public required IReadOnlyList<PokemonDto> FeaturedPokemon { get; init; }

        /// <summary>Redeemable promo codes associated with this event.</summary>
        [JsonPropertyName("promoCodes")]
        public required IReadOnlyList<string> PromoCodes { get; init; }
    }
}
